### PR TITLE
MLIBZ-2579: Codable Support

### PR DESCRIPTION
**Pull request by heyzooi**
#### Description

`Swift.Codable` Support

#### Changes

- Support for `Swift.Codable`
- Support for Custom JSON Parser
- `ObjectMapper` support deprecated

#### Tests

- Same Unit Tests (making sure ObjectMapper still working)
- Codable Unit Test
- Custom JSON Parser Unit Test

### CODE REVIEW

File: Kinvey/Kinvey/Acl.swift [hunk 1]
```diff
@@ -118,12 +89,61 @@ public final class Acl: Object, BuilderType {
...
+
+@available(*, deprecated: 3.18.0, message: "Please use Swift.Codable instead")
 extension Acl: Mappable {
     
     /// Constructor that validates if the map contains at least the creator.
     public convenience init?(map: Map) {
         var creator: String?
-        creator <- map[Acl.CodingKeys.creator]
+        creator <- (Acl.CodingKeys.creator.rawValue, map[Acl.CodingKeys.creator])
         if let creator = creator {
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/Acl.swift]: Enums in Swift can be `String` for example, so in this case a `rawValue` is actually the `String` value. For example `Acl.CodingKeys.creator` is actually a `case` of the `Acl.CodingKeys` enum, but if you need the string value wrapped in the enum you need to ask for the `rawValue`. Here's more details about it: https://developer.apple.com/documentation/swift/rawrepresentable

File: Kinvey/Kinvey/File.swift [hunk 2]
```diff
@@ -132,18 +132,19 @@ open class File: Object, Mappable {
...
+        metadata <- ("metadata", map[Entity.EntityCodingKeys.metadata])
+        publicAccessible <- ("publicAccessible", map["_public"])
+        fileName <- ("fileName", map["_filename"])
+        mimeType <- ("mimeType", map["mimeType"])
+        size.value <- ("size", map["size"])
+        upload <- ("upload", map["_uploadURL"])
+        download <- ("download", map["_downloadURL"])
+        expiresAt <- ("expiresAt", map["_expiresAt"], KinveyDateTransform())
+        uploadHeaders <- ("uploadHeaders", map["_requiredHeaders"])
     }
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/File.swift]: good point, let me fix that

File: Kinvey/Kinvey/FindOperation.swift [hunk 3]
```diff
@@ -186,8 +185,20 @@ internal class FindOperation<T: Persistable>: ReadOperation<T, AnyRandomAccessCo
...
         let startTime = CFAbsoluteTimeGetCurrent()
+        let client = options?.client ?? self.client
         let entities = AnyRandomAccessCollection(jsonArray.lazy.map { (json) -> T in
-            guard let entity = T(JSON: json, context: self.validationStrategy) else {
+            if let validationStrategy = self.validationStrategy {
+                do {
+                    try validationStrategy.validate(jsonArray: [json])
+                } catch {
+                    fatalError(error.localizedDescription)
+                }
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/FindOperation.swift]: that's actually the way to do a `try/catch` in Swift. If any of your calls inside a block can throw an error you need to wrap into a `do {}` block and then add `try` for each call that can throw an error

File: Kinvey/Kinvey/JSONParser.swift [hunk 4]
```diff
@@ -0,0 +1,368 @@
+//
+//  ResponseParser.swift
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/JSONParser.swift]: you are right, this file was renamed and i forgot to change the comments

File: Kinvey/Kinvey/ObjectMapperJSONParser.swift [hunk 5]
```diff
@@ -0,0 +1,116 @@
+//
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/ObjectMapperJSONParser.swift]: I decided to move all deprecated methods into one single file so can remove the file when we decided to remove support for ObjectMapper

File: Kinvey/Kinvey/Persistable.swift [hunk 6]
```diff
@@ -103,324 +80,6 @@ fileprivate func _kinveyMappingType(left: String, right: String, transform: AnyT
...
-            }
-            return nil
-        })
-    }
-    
-}
-
-/// Overload operator for `List` values
-public func <-<T: Object & BaseMappable>(lhs: List<T>, rhs: (String, Map)) {
-    let (right, map) = rhs
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/Persistable.swift]: no, the idea is you should use either `Codable` or `ObjectMapper`. Those operators should be only used if you still using `ObjectMapper`

File: Kinvey/Kinvey/Push.swift [hunk 7]
```diff
@@ -165,7 +165,7 @@ open class Push {
      }
      ```
      */
-    @available(iOS, deprecated: 10.0, message: "Please use registerForNotifications() instead")
+    @available(iOS, deprecated: 3.18.0, message: "Please use registerForNotifications() instead")
     open func registerForPush(
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/Push.swift]: this was a mistake, the deprecated version must be the version of your framework, but I would like to deprecate based in a version of the OS. Let me double check again to make sure there's no way to make this deprecation based in the OS.

**Comment by vinaygahlawat** [comment]
LGTM

**Pull request merged**